### PR TITLE
vtk: unvendor some of the dependencies, fix on darwin

### PR DESCRIPTION
**Pull request by veprbl**
###### Description of changes

This prevents vtk from building its own dependencies from scratch, which allows to fix it on darwin.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
Nice, I tried the same in https://github.com/NixOS/nixpkgs/pull/171934 but never managed to make it build.

**Comment by veprbl** [comment]
> Nice, I tried the same in #171934 but never managed to make it build.

Oh, would be nice to sometime rebase your work on top of this. I'd prefer to first review this as is first.

**Comment by veprbl** [comment]
@GrahamcOfBorg build vtk_7
@GrahamcOfBorg build vtk_9
@GrahamcOfBorg build python3Packages.vtk

**Comment by veprbl** [comment]
@GrahamcOfBorg f3d

**Comment by veprbl** [comment]
@GrahamcOfBorg build f3d

**Comment by bcdarwin** [comment]
The `f3d` build fails during configuring due to missing libraries but perhaps as with #178367 propagating some of these additional libraries from `vtk_9` is the correct solution.

**Comment by veprbl** [comment]
> The `f3d` build fails during configuring due to missing libraries but perhaps as with #178367 propagating some of these additional libraries from `vtk_9` is the correct solution.

Thank you for this summary. I agree, propagating the libraries is a solution. Although, it's a bit unusual that they are needed in runtime. One would think that VTK developers how to best use CMake, but it could be still a bug caused by lack of testing in environments with separated dependencies.

**Comment by bcdarwin** [comment]
Sorry, I meant configurePhase -- this is a build time, not runtime, failure.

**Comment by bcdarwin** [comment]
Maybe time to rebase this now that some other VTK-related issues are fixed?

File: pkgs/development/libraries/vtk/generic.nix [hunk 1]
```diff
@@ -65,13 +71,28 @@ in stdenv.mkDerivation rec {
...
+    "-D${if lib.versionOlder version "9.0" then "VTK_USE_SYSTEM_LIBHARU" else "VTK_MODULE_USE_EXTERNAL_VTK_libharu"}=OFF" # requires patches
     "-DOPENGL_INCLUDE_DIR=${libGL}/include"
     "-DCMAKE_INSTALL_LIBDIR=lib"
     "-DCMAKE_INSTALL_INCLUDEDIR=include"
     "-DCMAKE_INSTALL_BINDIR=bin"
     "-DVTK_VERSIONED_INSTALL=OFF"
+    "-DVTK_MODULE_USE_EXTERNAL_VTK_exprtk=OFF"
+    "-DVTK_MODULE_USE_EXTERNAL_VTK_cgns=OFF"
+    "-DVTK_MODULE_USE_EXTERNAL_VTK_ioss=OFF"
+  ] ++ optionals (lib.versionOlder version "8.0") [
```
Comment (bcdarwin) [inline pkgs/development/libraries/vtk/generic.nix:85]: VTK 7 has been dropped so this conditional isn't needed any longer.

File: pkgs/development/libraries/vtk/generic.nix [hunk 2]
```diff
@@ -65,13 +71,28 @@ in stdenv.mkDerivation rec {
...
+    "-DVTK_USE_SYSTEM_GL2PS=OFF" # needs an older version
+  ] ++ optionals (lib.versionOlder version "9.0") [
+    "-DVTK_USE_SYSTEM_LIBPROJ=OFF" # buggy include paths
+    "-DVTK_USE_SYSTEM_PUGIXML=OFF" # fails to detect
+    "-DVTK_USE_SYSTEM_OGG=OFF" # undefined symbols during linking
+    "-DVTK_USE_SYSTEM_THEORA=OFF" # undefined symbols during linking
+    "-DVTK_USE_SYSTEM_JSONCPP=OFF" # undefined symbols during linking
   ] ++ optionals enableQt [
     "-D${if lib.versionOlder version "9.0" then "VTK_Group_Qt:BOOL=ON" else "VTK_GROUP_ENABLE_Qt:STRING=YES"}"
   ] ++ optionals (enableQt && lib.versionOlder version "8.0") [
```
Comment (bcdarwin) [inline pkgs/development/libraries/vtk/generic.nix:98]: Same here.

**Comment by veprbl** [comment]
> Maybe time to rebase this now that some other VTK-related issues are fixed?

Would you like to give it a try?

**Pull request closed without merging**